notebooks/integrations/openai/function-calling.ipynb

```diff
@@ -0,0 +1,502 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "eac77545-c31c-4daa-bf5d-1a06f88ba8e8",
+   "metadata": {},
+   "source": [
+    "# OpenAI function calling with Elasticsearch"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "75f21771-cafe-496f-8aaf-7cc0f14f89ec",
+   "metadata": {},
+   "source": [
+    "[Function calling](https://platform.openai.com/docs/guides/function-calling) in OpenAI refers to the capability of AI models to interact with external functions or APIs, allowing them to perform tasks beyond text generation. This feature enables the model to execute code, retrieve information from databases, interact with external services, and more, by calling predefined functions.\n",
+    "\n",
+    "In this notebook we’re going to create two function:  \n",
+    "`fetch_from_elasticsearch()` - Fetch data from Elasticsearch using natural language query.   \n",
+    "`weather_report()` - Fetch a weather report for a particular location.\n",
+    "\n",
+    "We'll integrate function calling to dynamically determine which function to call based on the user's query and generate the necessary arguments accordingly."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "755e60d5-a42a-4001-a3a4-c5fe2da7d51a",
+   "metadata": {},
+   "source": [
+    "# Setup"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "12d5e15c-5f1d-462a-9b04-83a9d58ad28c",
+   "metadata": {
+    "jp-MarkdownHeadingCollapsed": true
+   },
+   "source": [
+    "### Elastic\n",
+    "\n",
+    "Create an [Elastic Cloud deployment](https://www.elastic.co/search-labs/tutorials/install-elasticsearch/elastic-cloud) to get all Elastic credentials.  \n",
+    "`ES_API_KEY`: [Create](https://www.elastic.co/search-labs/tutorials/install-elasticsearch/elastic-cloud#creating-an-api-key) an API key.  \n",
+    "`ES_ENDPOINT`: [Copy](https://www.elastic.co/search-labs/tutorials/install-elasticsearch/elastic-cloud#finding-your-cloud-id) endpoint of Elasticsearch.\n",
+    "\n",
+    "### Open AI\n",
+    "\n",
+    "`OPENAI_API_KEY`: Setup an [Open AI account and create a secret key](https://platform.openai.com/docs/quickstart).  \n",
+    "`GPT_MODEL`: We’re going to use the `gpt-4-1106-vision-preview` model but you can check [here](https://platform.openai.com/docs/guides/function-calling) which model is being supported for function calling.\n",
```

Comment:
worth using gpt-4o instead?

Reply:
Yeah I was about to change this. Updating with `gpt-4o`.